cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java

```diff
@@ -427,14 +446,39 @@ private void executeCondition(ApplicationSpecification appSpec, WorkflowConditio
     executeAll(iterator, appSpec, instantiator, classLoader, token);
   }
 
+  private void createLocalDatasets() throws IOException, DatasetManagementException {
+    for (Map.Entry<String, DatasetCreationSpec> instanceEntry : workflowSpec.getLocalDatasetSpecs().entrySet()) {
+      String localInstanceName = getLocalDatasetName(instanceEntry.getKey(), runId.getId());
+      Id.DatasetInstance instanceId = Id.DatasetInstance.from(program.getNamespaceId(), localInstanceName);
+      DatasetCreationSpec instanceSpec = instanceEntry.getValue();
+      LOG.info("Adding Workflow local dataset instance: {}", localInstanceName);
+      datasetFramework.addInstance(instanceSpec.getTypeName(), instanceId, instanceSpec.getProperties());
+    }
+  }
+
+  private void deleteLocalDatasets() {
+    for (Map.Entry<String, DatasetCreationSpec> instanceEntry : workflowSpec.getLocalDatasetSpecs().entrySet()) {
+      String localInstanceName = getLocalDatasetName(instanceEntry.getKey(), runId.getId());
+      Id.DatasetInstance instanceId = Id.DatasetInstance.from(program.getNamespaceId(), localInstanceName);
+      LOG.info("Deleting Workflow local dataset instance: {}", localInstanceName);
+      try {
+        datasetFramework.deleteInstance(instanceId);
+      } catch (IOException | DatasetManagementException e) {
+        LOG.warn("Failed to delete the Workflow local dataset instance {}", localInstanceName, e);
+      }
+    }
+  }
+
   @Override
   protected void run() throws Exception {
     LOG.info("Start workflow execution for {}", workflowSpec);
+    createLocalDatasets();
     WorkflowToken token = new BasicWorkflowToken(cConf.getInt(Constants.AppFabric.WORKFLOW_TOKEN_MAX_SIZE_MB));
     executeAll(workflowSpec.getNodes().iterator(), program.getApplicationSpecification(),
                new InstantiatorFactory(false), program.getClassLoader(), token);
 
     LOG.info("Workflow execution succeeded for {}", workflowSpec);
+    deleteLocalDatasets();
```

Comment:
So the contract is if a workflow execution fail, local dataset will never get deleted?